packages/eslint-plugin/src/rules/strict-enums.ts

```diff
@@ -0,0 +1,804 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+import * as tsutils from 'tsutils';
+import * as ts from 'typescript';
+
+import * as util from '../util';
+
+/**
+ * TypeScript only allows number enums, string enums, or mixed enums with both
+ * numbers and strings.
+ *
+ * Mixed enums are be a union of a number enum and a string enum, so there is
+ * no separate kind for them.
+ */
+enum EnumKind {
+  NON_ENUM,
+  HAS_NUMBER_VALUES,
+  HAS_STRING_VALUES,
+}
+
+/** These operators are always considered to be safe. */
+const ALLOWED_ENUM_OPERATORS = new Set(['in', '|', '&', '^', '|=', '&=', '^=']);
+
+/**
+ * See the comment for `EnumKind`.
+ *
+ * This rule can safely ignore other kinds of types (and leave the validation in
+ * question to the TypeScript compiler).
+ */
+const IMPOSSIBLE_ENUM_TYPES =
+  ts.TypeFlags.BooleanLike |
+  ts.TypeFlags.NonPrimitive |
+  ts.TypeFlags.ESSymbolLike;
+
+const ALLOWED_TYPES_FOR_ANY_ENUM_ARGUMENT =
+  ts.TypeFlags.Any |
+  ts.TypeFlags.Unknown |
+  ts.TypeFlags.Number |
+  ts.TypeFlags.String;
+
+export type Options = [];
+export type MessageIds =
+  | 'incorrectIncrement'
+  | 'mismatchedAssignment'
+  | 'mismatchedComparison'
+  | 'mismatchedFunctionArgument';
+
+export default util.createRule<Options, MessageIds>({
+  name: 'strict-enums',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Disallow the usage of unsafe enum code patterns',
+      recommended: false,
```

Comment:
We can do this now! 🎉 

```suggestion
      recommended: 'strict',
```